File: backend/app/governance/trace_authority.py
"""Authority tier labels for debug / governance trace panels (metadata only)."""

from __future__ import annotations

from typing import Any, Literal

AuthorityTier = Literal["AUTHORITATIVE", "PLANNING", "ADVISORY", "DIAGNOSTIC"]

TIER_AUTHORITATIVE: AuthorityTier = "AUTHORITATIVE"
TIER_PLANNING: AuthorityTier = "PLANNING"
TIER_ADVISORY: AuthorityTier = "ADVISORY"
TIER_DIAGNOSTIC: AuthorityTier = "DIAGNOSTIC"


def authority_label(
    tier: AuthorityTier,
    note: str,
) -> dict[str, str]:
    return {"authority_tier": tier, "authority_note": note}


def attach_authority_tier(
    payload: dict[str, Any] | None,
    *,
    tier: AuthorityTier,
    note: str,
) -> dict[str, Any] | None:
    if payload is None:
        return None
    enriched = dict(payload)
    enriched.update(authority_label(tier, note))
    return enriched


def build_control_plane_authority_index(
    *,
    has_run_contract: bool,
    has_final_evidence_gate: bool,
) -> dict[str, dict[str, str]]:
    index: dict[str, dict[str, str]] = {
        "route_adjudication": authority_label(
            TIER_AUTHORITATIVE,
            "Final route adjudication after deterministic policy.",
        ),
        "routing_provenance": authority_label(
            TIER_AUTHORITATIVE,
            "Canonical routing provenance for the live turn.",
        ),
        "evidence_plan": authority_label(
            TIER_PLANNING,
            "Evidence collection plan; planning snapshot only.",
        ),
        "resource_planner": authority_label(
            TIER_PLANNING,
            "Composed ResourcePlan decisions; not execution authority.",
        ),
        "planning_decision": authority_label(
            TIER_PLANNING,
            "Planner path selection metadata.",
        ),
        "llm_intent_advisory": authority_label(
            TIER_ADVISORY,
            "LLM intent advisory; deterministic policy wins on conflict.",
        ),
        "llm_advisory_trace": authority_label(
            TIER_ADVISORY,
            "LLM advisory attempt/candidate metadata; dropped reasons are not final failure.",
        ),
        "llm_plan_validation": authority_label(
            TIER_ADVISORY,
            "LLM plan validation outcome; advisory unless promoted by policy.",
        ),
        "rag_trace": authority_label(
            TIER_ADVISORY,
            "RAG retrieval hints and evidence refs.",
        ),
        "precondition_evaluation": authority_label(
            TIER_DIAGNOSTIC,
            "Route-plan shadow precondition evaluation (diagnostic).",
        ),
        "candidate_spl_generation": authority_label(
            TIER_PLANNING,
            "SPL candidate generation metadata prior to final RunContract projection.",
        ),
        "spl_slot_binding": authority_label(
            TIER_DIAGNOSTIC,
            "Slot binding validator diagnostics unless projected by RunContract block reason.",
        ),
        "mcp_execution": authority_label(
            TIER_AUTHORITATIVE,
            "Final MCP execution gate outcome for the turn when present.",
        ),
        "answer_contract": authority_label(
            TIER_AUTHORITATIVE,
            "Answer contract read-model for finalize.",
        ),
        "final_answer_validation": authority_label(
            TIER_AUTHORITATIVE,
            "Final answer validator outcome.",
        ),
    }
    if has_run_contract:
        index["run_contract"] = authority_label(
            TIER_AUTHORITATIVE,
            "RunContract owns final-run public posture: render, SPL lifecycle, MCP posture.",
        )
    if has_final_evidence_gate:
        index["final_evidence_gate"] = authority_label(
            TIER_AUTHORITATIVE,
            "FinalEvidenceGate owns evidence-derived HIL, live-language, and MITRE/severity caps.",
        )
    return index

File: frontend/src/lib/traceAuthority.ts
/** Diagnostic authority-tier labels from control_plane_trace (metadata only). */

export type AuthorityTier = 'AUTHORITATIVE' | 'PLANNING' | 'ADVISORY' | 'DIAGNOSTIC';

export interface AuthorityTierEntry {
  authority_tier: AuthorityTier;
  authority_note?: string;
}

export interface TraceAuthoritySectionRow {
  key: string;
  label: string;
  tier: AuthorityTier;
  note?: string;
  source: 'index' | 'section';
}

const TIER_ORDER: AuthorityTier[] = ['AUTHORITATIVE', 'PLANNING', 'ADVISORY', 'DIAGNOSTIC'];

export const TRACE_AUTHORITY_SECTION_LABELS: Record<string, string> = {
  route_adjudication: 'Route adjudication',
  routing_provenance: 'Routing provenance',
  evidence_plan: 'Evidence plan',
  resource_planner: 'Resource planner',
  planning_decision: 'Planning decision',
  llm_intent_advisory: 'LLM intent advisory',
  llm_advisory_trace: 'LLM advisory trace',
  llm_plan_validation: 'LLM plan validation',
  rag_trace: 'RAG trace',
  precondition_evaluation: 'Route-plan shadow preconditions',
  candidate_spl_generation: 'SPL candidate generation',
  spl_slot_binding: 'SPL slot binding',
  spl_artifact_handoff_summary: 'SPL artifact handoff',
  slot_constraint_projection: 'SPL slot constraint projection',
  mcp_execution: 'MCP execution gate',
  answer_contract: 'Answer contract',
  final_answer_validation: 'Final answer validation',
  run_contract: 'RunContract',
  final_evidence_gate: 'FinalEvidenceGate',
  route_plan_shadow_authority: 'Route-plan shadow',
};

const SECTION_SCAN_KEYS = [
  'run_contract',
  'final_evidence_gate',
  'evidence_plan',
  'resource_planner',
  'route_adjudication',
  'routing_provenance',
  'llm_advisory_trace',
  'llm_intent_advisory',
  'llm_plan_validation',
  'rag_trace',
  'candidate_spl_generation',
  'spl_artifact_handoff_summary',
  'spl_slot_binding',
  'slot_constraint_projection',
  'mcp_execution',
  'answer_contract',
  'final_answer_validation',
  'planning_decision',
  'precondition_evaluation',
  'route_plan_shadow_authority',
] as const;

function isAuthorityTier(value: unknown): value is AuthorityTier {
  return value === 'AUTHORITATIVE' || value === 'PLANNING' || value === 'ADVISORY' || value === 'DIAGNOSTIC';
}

function readAuthorityEntry(value: unknown): AuthorityTierEntry | null {
  if (!value || typeof value !== 'object') {
    return null;
  }
  const record = value as Record<string, unknown>;
  if (!isAuthorityTier(record.authority_tier)) {
    return null;
  }
  return {
    authority_tier: record.authority_tier,
    authority_note: typeof record.authority_note === 'string' ? record.authority_note : undefined,
  };
}

export function formatTraceAuthoritySectionLabel(key: string): string {
  return TRACE_AUTHORITY_SECTION_LABELS[key] ?? key.replace(/_/g, ' ');
}

export function readTraceAuthorityIndex(
  controlPlaneTrace: Record<string, unknown> | null | undefined,
): Record<string, AuthorityTierEntry> | null {
  const raw = controlPlaneTrace?.trace_authority_index;
  if (!raw || typeof raw !== 'object') {
    return null;
  }
  const index: Record<string, AuthorityTierEntry> = {};
  for (const [key, value] of Object.entries(raw as Record<string, unknown>)) {
    const entry = readAuthorityEntry(value);
    if (entry) {
      index[key] = entry;
    }
  }
  return Object.keys(index).length ? index : null;
}

export function collectTraceAuthorityRows(
  controlPlaneTrace: Record<string, unknown> | null | undefined,
): TraceAuthoritySectionRow[] {
  if (!controlPlaneTrace) {
    return [];
  }

  const rows: TraceAuthoritySectionRow[] = [];
  const seen = new Set<string>();

  const index = readTraceAuthorityIndex(controlPlaneTrace);
  if (index) {
    for (const [key, entry] of Object.entries(index)) {
      rows.push({
        key,
        label: formatTraceAuthoritySectionLabel(key),
        tier: entry.authority_tier,
        note: entry.authority_note,
        source: 'index',
      });
      seen.add(key);
    }
  }

  for (const key of SECTION_SCAN_KEYS) {
    if (seen.has(key)) {
      continue;
    }
    const entry = readAuthorityEntry(controlPlaneTrace[key]);
    if (!entry) {
      continue;
    }
    rows.push({
      key,
      label: formatTraceAuthoritySectionLabel(key),
      tier: entry.authority_tier,
      note: entry.authority_note,
      source: 'section',
    });
    seen.add(key);
  }

  rows.sort((left, right) => {
    const tierDelta = TIER_ORDER.indexOf(left.tier) - TIER_ORDER.indexOf(right.tier);
    if (tierDelta !== 0) {
      return tierDelta;
    }
    return left.label.localeCompare(right.label);
  });

  return rows;
}

export function hasTraceAuthorityData(
  controlPlaneTrace: Record<string, unknown> | null | undefined,
): boolean {
  return collectTraceAuthorityRows(controlPlaneTrace).length > 0;
}

export function summarizeTraceAuthorityTiers(rows: TraceAuthoritySectionRow[]): string {
  const counts = new Map<AuthorityTier, number>();
  for (const row of rows) {
    counts.set(row.tier, (counts.get(row.tier) ?? 0) + 1);
  }
  return TIER_ORDER.filter((tier) => counts.has(tier))
    .map((tier) => `${counts.get(tier)} ${tier.toLowerCase()}`)
    .join(' · ');
}
